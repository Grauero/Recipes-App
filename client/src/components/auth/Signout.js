import React from 'react';
import PropTypes from 'prop-types';
import { ApolloConsumer } from 'react-apollo';
import { withRouter } from 'react-router-dom';

function handleSignout(client, history) {
  localStorage.setItem('token', '');
  client.resetStore();
  history.push('/');
}

const Signout = ({ history }) => (
  <ApolloConsumer>
    {client => (
      <button type="button" onClick={() => handleSignout(client, history)}>
        Signout
      </button>
    )}
  </ApolloConsumer>
);

Signout.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

export default withRouter(Signout);
