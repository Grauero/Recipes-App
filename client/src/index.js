import React, { Fragment, lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import {
  BrowserRouter as Router, Route, Switch, Redirect
} from 'react-router-dom';
import ApolloClient from 'apollo-boost';
import { ApolloProvider } from 'react-apollo';
import * as serviceWorker from './serviceWorker';

import App from './components/App';
import Navbar from './components/layouts/Navbar';
import Spinner from './components/common/Spinner';
import withSession from './components/utils/withSession';
import './index.css';

const Search = lazy(() => import('./components/recipe/Search'));
const Signin = lazy(() => import('./components/auth/Signin'));
const Signup = lazy(() => import('./components/auth/Signup'));
const AddRecipe = lazy(() => import('./components/recipe/AddRecipe'));
const RecipePage = lazy(() => import('./components/recipe/RecipePage'));
const Profile = lazy(() => import('./components/profile/Profile'));

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_SERVER,
  fetchOptions: {
    credentials: true
  },
  request: (operation) => {
    const token = localStorage.getItem('token');

    operation.setContext({
      headers: {
        authorization: token
      }
    });
  },
  onError: ({ networkError }) => {
    if (networkError) {
      console.log('Network Error', networkError);
    }
  }
});

const Root = ({ refetch, session }) => (
  <Router>
    <Fragment>
      <Navbar session={session} />
      <Switch>
        <Route exact path="/" component={App} />
        <Route path="/search" component={Search} />
        <Route path="/signin" render={() => <Signin refetch={refetch} />} />
        <Route path="/signup" render={() => <Signup refetch={refetch} />} />
        <Route path="/recipe/add" render={() => <AddRecipe session={session} />} />
        <Route path="/recipes/:_id" component={RecipePage} />
        <Route path="/profile" render={() => <Profile session={session} />} />
        <Redirect to="/" />
      </Switch>
    </Fragment>
  </Router>
);

Root.propTypes = {
  refetch: PropTypes.func.isRequired,
  session: PropTypes.instanceOf(Object).isRequired
};

const RootWithSession = withSession(Root);

ReactDOM.render(
  <ApolloProvider client={client}>
    <Suspense fallback={<div className="fallback"><Spinner /></div>}>
      <RootWithSession />
    </Suspense>
  </ApolloProvider>,
  document.getElementById('root')
);

serviceWorker.unregister();
