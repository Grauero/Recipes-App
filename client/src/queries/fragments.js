import { gql } from 'apollo-boost';

const recipeFragments = {
  recipe: gql`
    fragment CompleteRecipe on Recipe {
      _id
      name
      imageUrl
      category
      description
      instructions
      createdDate
      likes
      username
    }
  `,
  like: gql`
    fragment LikeRecipe on Recipe {
      _id
      likes
    }
  `
};

export default recipeFragments;
