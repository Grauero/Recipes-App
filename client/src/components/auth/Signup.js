import React, { Component } from 'react';
import { Mutation } from 'react-apollo';

import Error from '../error/Error';
import { SIGNUP_USER } from '../../queries';

const initialState = {
  username: '',
  email: '',
  password: '',
  passwordConfirmation: ''
};

class Signup extends Component {
  state = { ...initialState };

  clearState = () => this.setState({ ...initialState });

  handleChange = e => this.setState({ [e.target.name]: e.target.value });

  handleSubmit = async (e, signupUser) => {
    e.preventDefault();

    const { data } = await signupUser();
    localStorage.setItem('token', data.signupUser.token);

    this.clearState();
  };

  validateForm = () => {
    const {
      username, email, password, passwordConfirmation
    } = this.state;
    const isInvalid = !username
      || !email
      || !password
      || !passwordConfirmation
      || password !== passwordConfirmation;

    return isInvalid;
  };

  render() {
    const {
      username, email, password, passwordConfirmation
    } = this.state;

    return (
      <div className="App">
        <h2 className="App">Signup</h2>
        <Mutation mutation={SIGNUP_USER} variables={{ username, email, password }}>
          {(signupUser, { loading, error }) => (
            <form className="form" onSubmit={e => this.handleSubmit(e, signupUser)}>
              <input
                type="text"
                name="username"
                placeholder="Username"
                value={username}
                onChange={this.handleChange}
              />

              <input
                type="email"
                name="email"
                placeholder="Email Address"
                value={email}
                onChange={this.handleChange}
              />

              <input
                type="password"
                name="password"
                placeholder="Password"
                value={password}
                onChange={this.handleChange}
              />

              <input
                type="password"
                name="passwordConfirmation"
                placeholder="Confirm Password"
                value={passwordConfirmation}
                onChange={this.handleChange}
              />

              <button
                type="submit"
                disabled={loading || this.validateForm()}
                className="button-primary"
              >
                Submit
              </button>

              {error && <Error error={error} />}
            </form>
          )}
        </Mutation>
      </div>
    );
  }
}

export default Signup;
