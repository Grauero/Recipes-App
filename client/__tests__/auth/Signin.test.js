import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';

import { Signin } from '../../src/components/auth/Signin';
import { SIGNIN_USER } from '../../src/queries';

const initialState = { username: 'username', password: 'password' };
const props = { history: { push: jest.fn() }, refetch: jest.fn() };

const mocks = [
  {
    request: {
      query: SIGNIN_USER,
      variables: {
        username: initialState.username,
        password: initialState.password
      }
    },
    result: {
      data: {
        signinUser: { token: 'token' }
      }
    }
  }
];
let component;
let wrapper;

beforeEach(() => {
  wrapper = mount(
    <MockedProvider mocks={mocks} addTypename={false}>
      <BrowserRouter>
        <Signin {...props} />
      </BrowserRouter>
    </MockedProvider>
  );

  component = wrapper.find(Signin);
  component.setState(initialState);
});

afterEach(() => {
  wrapper.unmount();
});

it('matches snapshot', () => {
  expect(toJSON(component)).toMatchSnapshot();
});

it('updates state while changing input value', () => {
  const input = component.find('input').get(0);
  input.props.onChange({ target: { name: 'username', value: 'updatedValue' } });

  expect(component.state()).not.toBe(initialState);
  expect(component.state().username).toBe('updatedValue');
});

it('calls the SIGNIN_USER mutation and writes auth token to localStorage when user submits form', async () => {
  component.find('form').simulate('submit');
  await wait(0);

  expect(window.localStorage.getItem('token')).toBe('token');
});

it('calls the SIGNIN_USER mutation and resets component state when user submits form', async () => {
  component.find('form').simulate('submit');
  await wait(0);

  expect(component.state()).not.toBe(initialState);
  expect(component.state().username).toBe('');
  expect(component.state().password).toBe('');
});

it('calls the SIGNIN_USER mutation and calls props.refetch() when user submits form', async () => {
  component.find('form').simulate('submit');
  await wait(0);

  expect(props.refetch).toHaveBeenCalled();
});

it('calls the SIGNIN_USER mutation and redirects to "/" when user submits form', async () => {
  component.find('form').simulate('submit');
  await wait(0);

  expect(props.history.push).toHaveBeenCalled();
  expect(props.history.push).toHaveBeenCalledWith('/');
});
