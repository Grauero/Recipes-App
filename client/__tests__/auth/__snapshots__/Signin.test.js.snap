// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Signin
  history={
    Object {
      "push": [Function],
    }
  }
  refetch={[Function]}
>
  <div
    className="App"
  >
    <h2
      className="App heading"
    >
      Signin
    </h2>
    <Mutation
      mutation={
        Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": undefined,
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "username",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "username",
                          },
                        },
                      },
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "password",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "password",
                          },
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "signinUser",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "token",
                          },
                          "selectionSet": undefined,
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [
                Object {
                  "defaultValue": undefined,
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NonNullType",
                    "type": Object {
                      "kind": "NamedType",
                      "name": Object {
                        "kind": "Name",
                        "value": "String",
                      },
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "username",
                    },
                  },
                },
                Object {
                  "defaultValue": undefined,
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NonNullType",
                    "type": Object {
                      "kind": "NamedType",
                      "name": Object {
                        "kind": "Name",
                        "value": "String",
                      },
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "password",
                    },
                  },
                },
              ],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 135,
            "start": 0,
          },
        }
      }
      variables={
        Object {
          "password": "",
          "username": "",
        }
      }
    >
      <form
        className="form"
        onSubmit={[Function]}
      >
        <input
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
          value=""
        />
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
        <button
          className="button-primary"
          disabled={true}
          type="submit"
        >
          Submit
        </button>
      </form>
    </Mutation>
  </div>
</Signin>
`;
